### DISCUSSION TITLE
Support more UNIX-like systems like FreeBSD and Solaris

### DISCUSSION THREAD

**Discussion by X547** [NVIDIA/open-gpu-kernel-modules#244] [Ideas]

**Comment by X547** [comment]
I am also interested in [Haiku operating system](https://www.haiku-os.org/) support that I use and contribute.

**Comment by Tfoedy** [comment]
I think that we should focus on Linux. The Dev team is smaller than for the Windows version of the driver so supporting multiple operating systems seems like a bad idea. Nobody really uses Haiku or any other different OS (for daily driving) than Linux/Windows/MacOS so we should focus on developing the best drivers we can for Linux.

**Reply by ryao** [reply]
So much of this is platform independent that keeping it Linux only does not give Linux much. Getting contributors from other platforms would help Linux more. The ZFSOnLinux driver gained FreeBSD and now it has active contributions from FreeBSD developers, which benefits Linux.

**Comment by extrowerk** [comment]
> Nobody really uses Haiku

Citation needed.

**Comment by kraileth** [comment]
> I think that we should focus on Linux. The Dev team is smaller than for the Windows version of the driver so supporting multiple operating systems seems like a bad idea.

On the contrary: Linux users have been arguing for the importance of such support e.g. by pointing to the fact that there's not just Windows and macOS. If the community would now rush to just be part of the "big three" and be ok with starving the rest, this is a betrayal of the ideal of Open Source. And this would hurt the whole ecosystem - including Linux.

> Nobody really uses Haiku or any other different OS (for daily driving) than Linux/Windows/MacOS so we should focus on developing the best drivers we can for Linux.

You are very wrong, sir (This comment was written on my FreeBSD workstation).

**Reply by modulok** [reply]
This comment was given a thumbs up from a FreeBSD workstation used for software development on a daily basis.

**Comment by Tfoedy** [comment]
>
> > Nobody really uses Haiku or any other different OS (for daily driving) than Linux/Windows/MacOS...
> 
> You are very wrong, sir (This comment was written on my FreeBSD workstation).

I respect anyone who uses Haiku or other Unix like systems. What I meant is that Windows, MacOS and Linux are mainstream, where mostly only tech geeks use other systems. I'm not saying there shouldn't be support in other systems, I'm just saying that we should mostly focus on polishing current drivers on Linux.

**Reply by ryao** [reply]
The Nvidia Linux driver is largely a subset of the Nvidia Windows driver, plus some compatibility shims. For everyone who uses the nvidia driver for Linux graphics, you have 100 people using it for Windows graphics. Rather than support Linux, why not mostly focus on polishing it for Windows?

Your logic can easily be used against Linux support.

**Comment by mtijanic** [comment]
Thank you for your interest!

Traditional NVIDIA drivers use the same binary blob (`nv-kernel.o`) on FreeBSD and Solaris and only have a minor OS-specific adapter layer that is available in source form when you download the drivers. The new "GSP" driver architecture is also very amenable to this approach and we do hope to support them eventually. 

Thus far, our focus has been on Linux only, and that is not likely to change for the next few releases while we add support for more consumer facing features (such as power management).

We will gladly work with the community if they decide to pursue this ahead of our own schedules, though. If anyone is planning on tackling this, I would strongly recommend reaching out to us first for guidance.

Thank you.

**Reply by despair86** [reply]
I posted my findings below, I take it that the dispatch table is populated in part or in whole by the (missing) firmware? If so, how could we move forward with making it available at boot?

**Comment by kraileth** [comment]
> I respect anyone who uses Haiku or other Unix like systems. What I meant is that Windows, MacOS and Linux are mainstream, where mostly only tech geeks use other systems. I'm not saying there shouldn't be support in other systems, I'm just saying that we should mostly focus on polishing current drivers on Linux.

Fair enough. There certainly is more demand for great Linux drivers, I also have no doubts there. Especially with efforts from Valve and Google's Stadia. But I wanted to bust that myth that "nobody" was seriously using other operating systems. I've come a long way from my first Riva TNT card (didn't like the 3dfx hype back then) when I still ran Windows to kind of fighting with nvidia cards on Linux to finally finding the proprietary driver in good shape on FreeBSD when I made that switch several years ago.

I don't have the time to play a lot of games anymore and am more interested in other features like supporting multiple screens. And going with nvidia has worked very well for me in that regard. While it's not an ideology for me, I do prefer open source drivers and would love to see other systems supported, too. However on platforms where leadership didn't try as hard to piss off everybody who doesn't love the GPL (there are very valid reasons for that) and nvidia specifically, things worked well before. It's great that things are moving into the right direction now and I hope that this marks the end of certain hostilities. But it would be very unfortunate if that lead to support for other operating systems to fall short. But according to mtijanic's answer that does not seem to be the plan which is great!

**Comment by owariee** [comment]
The driver is now open source, the FreeBSD guys can do the same thing they are doing for AMDGPU/Mesa drivers and just write a layer so the driver will run as expected. I think this is the best choice for the "not mainstream OSes" since this is what happens since the begin. Linux is growing(thanks VALVe/Community) and the win that linux acquired will benefit all the other open source OSes by no doubt.

**Comment by X547** [comment]
I hope that NVidia will keep current architecture with most driver part OS independent and per-OS support layers. That approach may be broken if NVidia decides to integrate driver to Linux upstream code base because Linux kernel maintainers may force to drop OS abstraction layer.

**Comment by shkhln** [comment]
"may" is an understatement. As a FreeBSD user I would be very pissed off about the termination of an (official) support.

>  if an approach similar to how the traditional drivers were handled would be feasible for bringing CUDA

[Tested with mixed results.](https://github.com/shkhln/libc6-shim/issues/1#issuecomment-643077457) Works for less ambitious things like OpenCL, though.

**Comment by shkhln** [comment]
Eh, that part refers to the post containing the word "may". Hence the quotation marks.

**Comment by dchmelik** [comment]
Great news about modules, but UNIX (*BSD/OpenSolaris/IllumOS--'genetic Unix') _is_ mainstream for scientists/programmers (and still networking, at a time was vast majority) because, as originally designed according to plan (instead of 'make up as you go along') has cleaner/stabler software such as kernels. IllumOS is the new Open Solaris, but I couldn't care less about plain Solaris nor other proprietary Unix (though I hope it and Sys7 to Sys10 to UnixWare become Free/Libre/Opensource Software (FLS, OSS, FOSS, FLOSS.))

**Comment by emaste** [comment]
> We will gladly work with the community if they decide to pursue this ahead of our own schedules, though. If anyone is planning on tackling this, I would strongly recommend reaching out to us first for guidance.

Hello from the [FreeBSD Foundation](https://freebsdfoundation.org/). We are very interested in seeing FreeBSD support in these open source drivers.

Although we don't have any plans to directly support this (with staff time) in the immediate future, we do entertain [project grant proposals](https://freebsdfoundation.org/get-involved/project-proposal-overview/) from the community. If someone is capable and interested in working on FreeBSD support in the near future please get in touch.

Foundation staff can also help with guidance and code review, either as part of a project grant or in collaboration with upstream.

**Comment by amonpaike** [comment]
> Nobody really uses Haiku or any other different OS

The funny thing is that I recognize in the comments at least 3 HAIKU excellent develpers full of potential,
One of these has already done the entire porting of HAIKU on RISC-V, he ported the AMD GPU drivers himself to run them on his RISC-V platform and since it was not enough, he ported the  x86 AMD gpu drivers  bypassing all the "junk" of DRI linux to collect better performance and better integration with the HAIKU philosophy, and directly supporting Vulkan and run Zink on top of it to have good OpenGL support and at the same time save coding and work.
I am sure he would give a big boost to the improvement of these nvidia drivers for all platforms..
I'm not a developer, but I'm a deep observer of this "no one used HAIKU"
I tell you just one thing, this HAIKU used by no one, is full of pioneers who look far ...
do not be obtuse.

**Comment by despair86** [comment]
I've managed to build this driver in two parts, to drive a GTX 1650 on Solaris 11, except that the current kmod source (same version) has no way of loading the GSP firmware with this driver: https://github.com/despair86/open-gpu-kernel-modules/blob/main/sunos-kmod/nvidia_subr.c

this is probably the source of a NULL pointer reference in `gpuPostConstruct_IMPL()`: 
```
panicstack = unix:die+c6 () | unix:trap+1169 () | unix:_cmntrap+e9 () | 0 () | nvidia:gpuPostConstruct_IMPL+135 () | nvidia:gpumgrAttachGpu+1d6 () | nvidia:osInitNvMapping+3ee () | nvidia:RmInitAdapter+31d () | nvidia:rm_init_adapter+a9 () | nvidia:nv_open_increment+41 () | nvidia:nv_open+491 () | genunix:dev_open+26 () | specfs:spec_open+1ab () | genunix:fop_open+96 () | genunix:vn_openat+20c () | genunix:copen+421 () | genunix:openat+29 () | genunix:open+1c () | unix:brand_sys_syscall+1fe ()
```
in spite of this, the proprietary driver does not appear to utilise GSP at all: `GSP Firmware Version                  : N/A` even on this `Turing` core

it tries to call `gpuClearFbhubPoisonIntrForBug2924523_HAL()` which points to nothing

**Reply by ryao** [reply]
I suggest making an issue to ask for help with that.

**Comment by X547** [comment]
I finally managed to get this driver working on Haiku with Mesa NVK Vulkan driver. It is still early work and many applications are not working properly.

No libdrm stuff is used.

![screenshot695](https://github.com/user-attachments/assets/504b80ea-1cd0-4c83-885f-a7db3bcd01b7)

**Reply by VVD** [reply]
Can you run `glxgears`?

**Reply by X547** [reply]
![screenshot696](https://github.com/user-attachments/assets/bf7a0123-3c49-42a7-9cf4-1c27a5da95c4)

**Reply by VVD** [reply]
Can it work without Vulkan support in graphics card?

**Reply by X547** [reply]
OpenGL can be used with Zink. For pure OpenGL Mesa Nouveau is the only option, but I not tried to port it. Zink+NVK should be good enough.

Nvidia kernel driver supports only Turing+ GPUs, so all of them supports Vulkan.

**Reply by VVD** [reply]
Heh, Tesla…

**Comment by X547** [comment]
Driver running on Haiku and RISC-V machine (HiFive Unmatched). This machine firmware can't initialize GPU at boot time so modesetting is performed by OS.

<img width="1920" height="1080" alt="screenshot1" src="https://github.com/user-attachments/assets/5d5f57d1-badd-431f-93fe-ea8aacba732c" />
